Fix #5315 - OpenStudio CLI unable to require uuid in OpenStudio 3.9 (thereby breaking dependencies like openstudio-extension)

jmarrec opened the pull request:
Pull request overview
---------------------

 - Fixes #5315
 - Windows only

### Pull Request Author

<!--- Add to this list or remove from it as applicable.  This is a simple templated set of guidelines. -->

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

ci-commercialbuildings commented:
CI Results for f703ea40e43be258e7592d5e34dab4c9443fc957:

* Ubuntu 24.04:
    * [OpenStudio-3.9.1-alpha+f703ea40e4-Ubuntu-24.04-x86_64.deb](http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-5316/OpenStudio-3.9.1-alpha%2Bf703ea40e4-Ubuntu-24.04-x86_64.deb)
    * [OpenStudio-3.9.1-alpha+f703ea40e4-Ubuntu-24.04-x86_64.tar.gz](http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-5316/OpenStudio-3.9.1-alpha%2Bf703ea40e4-Ubuntu-24.04-x86_64.tar.gz)

* Ubuntu 22.04:
* Darwin:
* Windows:
* Code Coverage:

jmarrec commented on src/cli/CMakeLists.txt:
```diff
@@ -203,6 +203,11 @@ if(BUILD_TESTING)
     PASS_REGULAR_EXPRESSION "Hello from -x, at National Renewable Energy Laboratory"
   )
 
+  # Test for 5315
+  add_test(NAME OpenStudioCLI.EmbeddedScripting.uuid
+    COMMAND $<TARGET_FILE:openstudio> ${CMAKE_CURRENT_SOURCE_DIR}/test/ensure_uuid_can_be_loaded.rb
+  )
```
Before fix on on Windows:

```
    Start 3757: OpenStudioCLI.EmbeddedScripting.uuid

3757: Test command: C:\src\OpenStudio\build-release\Products\openstudio.exe "C:/src/OpenStudio/src/cli/test/ensure_uuid_can_be_loaded.rb"
3757: Working Directory: C:/src/OpenStudio/build-release/src/cli
3757: Test timeout computed to be: 1500
3757:
3757: NoMethodError: undefined method `flock' for #<StringIO:0x0000021f46ae7380>
3757: Backtrace:
3757:   :/ruby/3.2.0/rubygems.rb:788:in `block in open_file'
3757:   eval:464:in `open'
3757:   :/ruby/3.2.0/rubygems.rb:785:in `open_file'
3757:   :/ruby/3.2.0/rubygems/specification.rb:1158:in `load'
3757:   :/ruby/3.2.0/gems/uuid-2.3.9/lib/uuid.rb:66:in `<module:Version>'
3757:   :/ruby/3.2.0/gems/uuid-2.3.9/lib/uuid.rb:65:in `<class:UUID>'
3757:   :/ruby/3.2.0/gems/uuid-2.3.9/lib/uuid.rb:62:in `<main>'
3757:   eval:236:in `eval'
3757:   eval:236:in `require_embedded_absolute'
3757:   eval:221:in `block in require_embedded'
3757:   eval:215:in `each'
3757:   eval:215:in `require_embedded'
3757:   eval:190:in `rescue in require'
3757:   eval:184:in `require'
3757:   <internal::/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
3757:   C:/src/OpenStudio/src/cli/test/ensure_uuid_can_be_loaded.rb:1:in `<top (required)>'
3757:   eval:182:in `require'
3757:   eval:182:in `require'
3757:   <internal::/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
3757:   eval:4:in `<main>'
3757: Failed to execute 'C:/src/OpenStudio/src/cli/test/ensure_uuid_can_be_loaded.rb'
1/1 Test #3757: OpenStudioCLI.EmbeddedScripting.uuid ...***Failed    2.67 sec
```

jmarrec commented on ruby/engine/embedded_help.rb:
```diff
@@ -56,6 +58,17 @@ def self.preprocess_ruby_script(s)
 # DLM: ignore for now
 #Encoding.default_external = Encoding::ASCII
 
+# We patch Kernel.open, and IO.open
+# to read embedded files as a StringIO, not as a FileIO
+# But Gem.open_file for eg expects it to be a File, not a StringIO, and on
+# Windows it will call File::flock (file lock) to protect access, but StringIO
+# does not have this method
+class FakeFileAsStringIO < StringIO
+  def flock
+    return 0
+  end
+end
```
In Kernel.open and IO.open, when we returned a StringIO, we now return this FakeFileAsStringIO which will have the flock method...

ci-commercialbuildings commented:
CI Results for 0fdf34a27a48ef171652b712e04f33c2c538ed5a:

* Ubuntu 24.04:
    * [OpenStudio-3.9.1-alpha+0fdf34a27a-Ubuntu-24.04-x86_64.deb](http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-5316/OpenStudio-3.9.1-alpha%2B0fdf34a27a-Ubuntu-24.04-x86_64.deb)
    * [OpenStudio-3.9.1-alpha+0fdf34a27a-Ubuntu-24.04-x86_64.tar.gz](http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-5316/OpenStudio-3.9.1-alpha%2B0fdf34a27a-Ubuntu-24.04-x86_64.tar.gz)

* Ubuntu 22.04:
* Darwin:
* Windows:
* Code Coverage:

jmarrec commented:
@wenyikuang concerned that the Windows hapilly reports green, when in reality tests have failed here.

ci-commercialbuildings commented:
CI Results for d3822140a93a20e3247dca71b1d7ad29aecef6cd:

* Ubuntu 24.04:
    * [OpenStudio-3.9.1-alpha+d3822140a9-Ubuntu-24.04-x86_64.deb](http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-5316/OpenStudio-3.9.1-alpha%2Bd3822140a9-Ubuntu-24.04-x86_64.deb)
    * [OpenStudio-3.9.1-alpha+d3822140a9-Ubuntu-24.04-x86_64.tar.gz](http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-5316/OpenStudio-3.9.1-alpha%2Bd3822140a9-Ubuntu-24.04-x86_64.tar.gz)

* Ubuntu 22.04:
* Darwin:
* Windows:
* Code Coverage:

ci-commercialbuildings commented:
CI Results for 91e9b4a5f22ee2ffd98f5ec1c4fd069af5b3e381:

* Ubuntu 20.04:
* Ubuntu 22.04:
* Darwin:
    * [OpenStudio-3.9.1-alpha+91e9b4a5f2-Darwin-x86_64.dmg](http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-5316/OpenStudio-3.9.1-alpha+91e9b4a5f2-Darwin-x86_64.dmg)
    * [OpenStudio-3.9.1-alpha+91e9b4a5f2-Darwin-x86_64.tar.gz](http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-5316/OpenStudio-3.9.1-alpha+91e9b4a5f2-Darwin-x86_64.tar.gz)

* Windows:
* Code Coverage:

ci-commercialbuildings commented:
CI Results for d3822140a93a20e3247dca71b1d7ad29aecef6cd:

* Ubuntu 24.04:
* Ubuntu 22.04:
    * [OpenStudio-3.9.1-alpha+d3822140a9-Ubuntu-22.04-x86_64.deb](http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-5316/OpenStudio-3.9.1-alpha%2Bd3822140a9-Ubuntu-22.04-x86_64.deb)
    * [OpenStudio-3.9.1-alpha+d3822140a9-Ubuntu-22.04-x86_64.tar.gz](http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-5316/OpenStudio-3.9.1-alpha%2Bd3822140a9-Ubuntu-22.04-x86_64.tar.gz)

* Darwin:
* Windows:
* Code Coverage:

wenyikuang commented:
You mean these ones:

```
The following tests FAILED:

	685 - BCLFixture.BCLMeasure_CTor_throw_invalid_xml (Failed)

	3801 - OpenStudioCLI.Run_AlfalfaWorkflow (Failed)

	3945 - CLITest-test_bundle-bundle (Failed)

	3946 - CLITest-test_bundle-bundle_git (Failed)

	3948 - CLITest-test_bundle-bundle_native_embedded (Failed)

	3950 - CLITest-test_bundle-no_bundle (Failed)

	3951 - CLITest-test_bundle-bundle_default (Failed)

	3987 - CLITest-test_encodings-encoding_external (Failed)

```
Right?

jmarrec commented:
Way more than these on this commit:
https://ci.openstudio.net/blue/organizations/jenkins/openstudio-incremental-windows/detail/PR-5316/2/

The test I just added failed. All of workflow based ones have been failing silently since before 3.9.0 (I fixed the add_dll_directory on this PR after realizing).

```
The following tests FAILED:

	685 - BCLFixture.BCLMeasure_CTor_throw_invalid_xml (Failed)

	3772 - OpenStudioTest.LoggerGlobal (Failed)

	3778 - OpenStudioCLI.python_version (Failed)

	3784 - OpenStudioCLI.python_execute_line (Exit code 0xc0000409

)

	3789 - OpenStudioCLI.execute_python_script (Failed)

	3790 - OpenStudioCLI.execute_python_script.forward_flags (Failed)

	3791 - OpenStudioCLI.execute_python_script.forward_flags.omit_execute (Failed)

	3792 - OpenStudioCLI.EmbeddedScripting.uuid (Failed)

	3797 - OpenStudioCLI.Run_PythonOnly (Failed)

	3798 - OpenStudioCLI.Run_PythonOnly_TwoMeasure (Failed)

	3799 - OpenStudioCLI.Run_RubyPython (Failed)

	3800 - OpenStudioCLI.Run_PythonRuby (Failed)

	3801 - OpenStudioCLI.Run_AlfalfaWorkflow (Failed)

	3802 - OpenStudioCLI.Run_RubyPythonPlugin (Failed)

	3811 - OpenStudioCLI.test_logger_py (Failed)

	3817 - OpenStudioCLI.execute_python_script.forward_flags.forward_help (Failed)

	3823 - OpenStudioCLI.execute_python_script.forward_flags.path_forwardslash (Failed)

	3824 - OpenStudioCLI.execute_python_script.forward_flags.path_backwardslash (Failed)

	3825 - OpenStudioCLI.execute_python_script.numpy.explicit_sys_path_insert (Failed)

	3826 - OpenStudioCLI.execute_python_script.numpy.python_path (Failed)

	3827 - OpenStudioCLI.execute_python_script.numpy.python_home (Failed)

	3945 - CLITest-test_bundle-bundle (Failed)

	3946 - CLITest-test_bundle-bundle_git (Failed)

	3948 - CLITest-test_bundle-bundle_native_embedded (Failed)

	3950 - CLITest-test_bundle-no_bundle (Failed)

	3951 - CLITest-test_bundle-bundle_default (Failed)

	3987 - CLITest-test_encodings-encoding_external (Failed)

	3988 - PythonEngineFixture.BadMeasure (Failed)

	3989 - PythonEngineFixture.WrongMethodMeasure (Failed)

	3990 - PythonEngineFixture.StackLevelTooDeepMeasure (Failed)

	3991 - PythonEngineFixture.AlfalfaMeasure (Failed)

```
